### PR TITLE
feat: Full/Improved pipeline

### PR DESCRIPTION
**Pull request by HarryET**
# Description
Full pipeline for Echo Server, uses followings steps now:
- Every main commit publishes an image to ECR with the commit hash
- Terraform is redeployed with the latest hash

If a public release is needed (for external parties), manually run the "release" action, which will:
- Bump version
- Generate changelog
- Publish release
- Publish images to ghcr & ECR with the version as tag
- Update infra to use versioned container

## How Has This Been Tested?
Actions can't be tested locally

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

**Comment by chris13524** [comment]
Is this ready for review? It is marked as draft.

PR description says:
> Every main commit publishes an image to ECR with the commit hash
> Terraform is redeployed with the latest hash

But what makes sure the Terraform is deployed after the ECR image is pushed?

**Comment by HarryET** [comment]
It's a draft because it is not ready

File: .github/workflows/publish_image.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
+name: publish_image
+
+on:
```
Comment (chris13524) [inline .github/workflows/publish_image.yml:3]: PR description says:

> Every main commit publishes an image to ECR with the commit hash
> Terraform is redeployed with the latest hash

But what makes sure the Terraform is deployed after the ECR image is pushed?

Comment (HarryET) [reply .github/workflows/publish_image.yml:3]: it has now been added

File: .github/workflows/auto_deploy.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      matrix:
+        file: [Dockerfile, slim.Dockerfile]
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      # Ensure that we get the new version from updated Cargo.toml
+      - name: Move to HEAD
+        run: |
+          git reset --hard HEAD
```
Comment (chris13524) [inline .github/workflows/auto_deploy.yml]: Isn't this redundant? actions/checkout should always checkout the latest commit

Comment (HarryET) [reply .github/workflows/auto_deploy.yml]: Yeah! Its redundent here, it was from another action where it's needed

**Review by arein — CHANGES_REQUESTED** [review]
you're deploying without running tests?

Suggestion: make this a step in `ci.yaml` that runs after the tests and only if `push` to `main`

**Comment by HarryET** [comment]
> you're deploying without running tests?
> 
> Suggestion: make this a step in `ci.yaml` that runs after the tests and only if `push` to `main`

Now runs CI before it deploys and `auto_deploy` only runs on commits to main

**Review by arein — CHANGES_REQUESTED** [review]
bit confused here? My previous feedback which you said was good e.g. the lack of running tests has been ignored here?

File: .github/workflows/auto_deploy.yml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        uses: docker/build-push-action@v3
+        with:
+          context: .
+          file: ${{ matrix.file }}
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+  run-ci:
```
Comment (HarryET) [inline .github/workflows/auto_deploy.yml:76]: @arein Tests are run here!

**Comment by HarryET** [comment]
> bit confused here? My previous feedback which you said was good e.g. the lack of running tests has been ignored here?

Tests are run on line 76, I tagged you in a comment :)

File: .github/workflows/auto_deploy.yml (hunk 3 above)
Comment (arein) [reply .github/workflows/auto_deploy.yml:76]: missed it. Thank you!

**Review by arein — APPROVED** [review]

**Pull request merged**